fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -1073,6 +1064,35 @@ public void rebuildIndex(@Nonnull FDBRecordStore store) {
         });
     }
 
+    private <T> CompletableFuture<T> runWithSynchronizedRunnerAndEndSession(
+            @Nonnull SynchronizedSessionRunner newSynchronizedRunner, @Nonnull Supplier<CompletableFuture<T>> runnable) {
+        if (this.synchronizedSessionRunner == null) {
+            this.synchronizedSessionRunner = newSynchronizedRunner;
+            return AsyncUtil.composeHandle(runnable.get(), (result, ex) -> {
+                if (this.synchronizedSessionRunner == newSynchronizedRunner) {
+                    this.synchronizedSessionRunner = null;
+                } else {
+                    LOGGER.warn(KeyValueLogMessage.of("synchronizedSessionRunner was modified during the run",
+                            LogMessageKeys.SESSION_ID, newSynchronizedRunner.getSessionId(),
+                            LogMessageKeys.INDEXER_SESSION_ID, this.synchronizedSessionRunner == null ? null : this.synchronizedSessionRunner.getSessionId()));
+                }
+                return newSynchronizedRunner.endSessionAsync().thenApply(vigore -> {
+                    if (ex == null) {
+                        return result;
+                    } else {
+                        throw new RuntimeException(ex);
+                    }
+                });
+            });
+        } else {
+            return newSynchronizedRunner.endSessionAsync().thenApply(vignore -> {
+                throw new RecordCoreException("another synchronized session is running on the indexer",
+                        LogMessageKeys.SESSION_ID, newSynchronizedRunner.getSessionId(),
+                        LogMessageKeys.INDEXER_SESSION_ID, this.synchronizedSessionRunner.getSessionId());
```

Comment:
It might actually make sense to preserve the object's synchronized session runner in a local variable, as I think there's a race condition here where:

1. The object's `syncrhonizedSessionRunner` is not null
1. Someone starts a second run
1. The new synchronized session runner ends it's session but meanwhile the initial runner completes and the object's variable is set to `null`
1. Then an NPE on this line

Reply:
Fixed.